benchmark/rocblas/utils.hpp

```diff
@@ -0,0 +1,206 @@
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename utils.hpp
+ *
+ **************************************************************************/
+
+#ifndef ROCBLAS_UTILS_HPP
+#define ROCBLAS_UTILS_HPP
+
+#include "sycl_blas.h"
+#include <common/common_utils.hpp>
+
+#include <hip/hip_runtime.h>
+#include <rocblas.h>
+
+#ifndef CHECK_HIP_ERROR
+#define CHECK_HIP_ERROR(error)                                    \
+  if (error != hipSuccess) {                                      \
+    fprintf(stderr, "hip error: '%s'(%d) at %s:%d\n",             \
+            hipGetErrorString(error), error, __FILE__, __LINE__); \
+  }
+#endif
+
+#ifndef CHECK_ROCBLAS_STATUS
+#define CHECK_ROCBLAS_STATUS(status)                                       \
+  if (status != rocblas_status_success) {                                  \
+    fprintf(stderr, "rocBLAS error: ");                                    \
+    fprintf(stderr, "rocBLAS error: '%s'(%d) at %s:%d\n",                  \
+            rocblas_status_to_string(status), status, __FILE__, __LINE__); \
+  }
+#endif
+
+namespace blas_benchmark {
+
+void create_benchmark(blas_benchmark::Args& args, rocblas_handle& rb_handle_ptr,
+                      bool* success);
+
+namespace utils {
+
+/**
+ * @class HIPDeviceMemory
+ * @brief Base-class to allocate/deallocate hip device memory.
+ * @tparam T is the type of the underlying data
+ * @tparam CopyToHost whether to copy back
+ */
+template <typename T, bool CopyToHost = false>
+class HIPDeviceMemory {
+ protected:
+  size_t size_;
+  size_t bytes_;
+
+  HIPDeviceMemory() : size_(1), bytes_(sizeof(T)) {}
+
+  HIPDeviceMemory(size_t s) : size_(s), bytes_(s * sizeof(T)) {}
+
+  // Allocate on Device
+  T* alloc() {
+    T* d;
+    if ((hipMalloc)(&d, bytes_) != hipSuccess) {
+      fprintf(stderr, "Error allocating %zu bytes_\n", bytes_);
+      d = nullptr;
+    }
+    return d;
+  }
+
+  // Copy from Host to Device
+  void copyH2D(const T* hPtr, T* dPtr) {
+    if (dPtr != nullptr) {
+      CHECK_HIP_ERROR(hipMemcpy(dPtr, hPtr, bytes_, hipMemcpyHostToDevice));
+    }
+    return;
+  }
+
+  // Copy from Device to Host
+  void copyD2H(const T* dPtr, T* hPtr) {
+    if (hPtr != nullptr) {
+      CHECK_HIP_ERROR(hipMemcpy(hPtr, dPtr, bytes_, hipMemcpyDeviceToHost));
+    }
+    return;
+  }
+
+  // Free device memory
+  void free(T* d) {
+    if (d != nullptr) {
+      CHECK_HIP_ERROR((hipFree)(d));
```

Comment:
I don't know much of hipAPI but this ```((hipFree)(d))``` looks different from other hipAPI calls, is it correct?